sphinx/themes/basic/static/searchtools.js

```diff
@@ -64,9 +72,13 @@ const _displayItem = (item, searchTerms, highlightTerms) => {
   const showSearchSummary = DOCUMENTATION_OPTIONS.SHOW_SEARCH_SUMMARY;
   const contentRoot = document.documentElement.dataset.content_root;
 
-  const [docName, title, anchor, descr, score, _filename] = item;
+  const [docName, title, anchor, descr, score, _filename, resultKind] = item;
 
   let listItem = document.createElement("li");
+  // Add a class representing the item's type:
+  // can be used by a theme's CSS selector for styling
+  // See SearchResultKind for the class names.
+  listItem.classList.add(`result-kind-${resultKind}`);
```

Comment:
@AA-Turner in https://github.com/sphinx-doc/sphinx/pull/12542 we don't seem to add any prefixes to domain/object-type when those are added as CSS classes.  What's the reason for adding a prefix here?

It feels to me like the CSS class names we choose here could be long-lived; this is the only part of the branch as-is that I feel cautious about.

Reply:
I had originally not included any prefix, because the typical selector will be `ul.search li.[type]` and `ul.search li.object` seemed clear enough. But I don’t have a strong opinion.

Reply:
Gradually iterating through some thoughts:

 * From the perspective of a theme author, the ability to reference consistent CSS names when styling search results and ToC entries seems like it'd be beneficial.
 * Sphinx-style object qualifiers (`py:module`, `cpp:class`) would conflict with CSS syntax - so `class="py module"` or `class="cpp class"` might be the preferred output here.
 * Building upon those two thoughts, the `kind` value that we're talking about here is useful context -- but it is a different dimension to the Sphinx object qualifier.  The values could conflict for some domains though -- so in fact, a prefix does seem like a good idea.

Reply:
One more thought: for some reason I don't love the word _kind_ in this context.  It seems like an unnecessarily-different synonym for `type` -- itself a frequently-overused name in software.

I'd like to express something about what the currently-named-`kind` here relates to.  And I think it's something like a _match category_ -- it's _how_ and/or _why_ the search algorithm decided to include an entry (yes, that's generally derived 1-1 from how the search index is structured, but it doesn't necessarily have to be).

Reply:
Prototyping / pseudogenerating some results:

```
<result class="match-object cpp class">utils::Format</result>
<result class="match-text">Formatting Guidelines</result>
<result class="match-object py function">foo.bar.format</result>
```

Does that seem semantic and themeable?

Reply:
(I'm _not_ suggesting that we include Sphinx domain and object type yet -- but trying to anticipate that they could be included in future)

Reply:
`class="py module"` seems a good extension possibility. I’ve first thought about `class="py-module"` because you likely want to be domain-specific and not style all modules the same way across contexts. However, you can easily combine selectors `.py.module`, which is not worse than `.py-module` and leaves more flexibility.

I’m not having a good prefix either. Personal thoughts: `type` is very generic. Same with `kind`, but that’s additionally feels somewhat unconventional. I`match` connects too much to the search process and less to the result. Other possibilities `context`, `category`

Not necessarily related but possibly a good way to come up with suitable names(prefix and type/kind): If we’d add a GitHub-like filter mechanism to the search field e.g.`some_func type:object`, what would be intuitive names for our end users?

Reply:
> Not necessarily related but possibly a good way to come up with suitable names(prefix and type/kind): If we’d add a GitHub-like filter mechanism to the search field, what would be intuitive names for our end users? e.g. `some_func category:api` feels better than`some_func kind:object`.

Mm, good question - query language syntax.  The `label:value` syntax does seem well established for various web-based search engines.  For that specific example, I think that `name:some_func` might be simplest - the function is an object entity that is indexed, and those entities tend to have names.  Other entities could be `document`, `section`, `word`... but I'd also be wary of reinventing the `docutils` node hierarchy.

Reply:
> It feels to me like the CSS class names we choose here could be long-lived; this is the only part of the branch as-is that I feel cautious about.

> I had originally not included any prefix, because the typical selector will be `ul.search li.[type]` and `ul.search li.object` seemed clear enough.

I added a prefix out of caution -- CSS backwards-compatibility is very annoying, as you've no idea what selectors people can use. Whilst ``ul.search li.[type]`` is *best*, ``li.[type]`` or even ``.[type]`` is *possible*, and hence reusing the class names somewhere else carries a small risk. Namespacing with some prefix reduces that. I haven't merged #12542 as I'm unsure if the benefit is worth the long-term backward compatibilty commitment. On the other hand, I might be overthinking this whole endevaour entirely.

> One more thought: for some reason I don't love the word _kind_ in this context. It seems like an unnecessarily-different synonym for `type` -- itself a frequently-overused name in software.

*type* felt wrong here, though on a 'feelings' level rather than a great deal of rationality. No strong feelings on going back to *type* really.

A

Reply:
Having pondered the options, I suggest to go with a prefix `category` or `context`. Both seem equally fine for me. I also suggest to change `object` to `api because not every item is an object.

Reply:
I'd like to suggest reserving `category` as a potentially-useful name for some other search-related feature (TBD).

Using `context` seems good to me.  "What context is the search result in?"  "It's in the context of a Python class, a C++ struct, ..."

I'm not so keen on `api` though.  Integer constants and string attributes, for example, don't seem like APIs to me.  They are 'domain objects' though, even if they're what Python terminology refers to as an `object`.  Also, I can theoretically imagine Sphinx domains that aren't programming languages, but do have some kind of multi-level entity structuring.

Reply:
Also a low-priority argument from my own laziness: I'd like to be able to ask people (with any range of knowledge) "what kind of results did you get?" when they mention a problem with search results, without them potentially thinking that I'm referring to the... whatever we're calling it... of the individual results they saw.

Reply:
> I'm not so keen on `api` though. Integer constants and string attributes, for example, don't seem like APIs to me. They are 'domain objects' though, even if they're not what Python terminology refers to as an `object`. Also, I can theoretically imagine Sphinx domains that aren't programming languages, but do have some kind of multi-level entity structuring.

Bringing that back to reality: the Sphinx [Domains documentation](https://github.com/sphinx-doc/sphinx/blob/de15d61a46daaaa2b0a0e341cdb4e0abe107e012/doc/usage/domains/index.rst) does state that it is intended to model programming language domains, and not others.

The Sphinx [index definition of `object`](https://www.sphinx-doc.org/en/master/glossary.html#term-object) is currently:

> The basic building block of Sphinx documentation. Every “object directive” (e.g. [py:function](https://www.sphinx-doc.org/en/master/usage/domains/python.html#directive-py-function) or [object](https://www.sphinx-doc.org/en/master/usage/domains/standard.html#directive-object)) creates such a block; and most objects can be cross-referenced to.

(sorry for being verbose)